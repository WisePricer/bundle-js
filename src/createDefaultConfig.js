import { resolve } from 'path'
import WebpackNotifierPlugin from 'webpack-notifier'
import { DefinePlugin } from 'webpack'
import TerserPlugin from 'terser-webpack-plugin'
import LiveReloadPlugin from 'webpack-livereload-plugin'
import CompressionPlugin from 'compression-webpack-plugin'
import HtmlPlugin from 'html-webpack-plugin'
import flatten from 'lodash.flatten'
import mapValues from 'lodash.mapvalues'

import { getGitCommit } from './getGitCommit'

const MODE = {
  DEVELOPMENT: 'development',
  PRODUCTION: 'production'
}

export const createDefaultConfig = ({
  mode = MODE.DEVELOPMENT,
  context = resolve('./src'),
  entry = './index.js',
  rules = [],
  output = {
    path: resolve('./dist')
  },
  html = true,
  htmlOptions = {},
  less = true,
  lessOptions = {},
  watch,
  watchOptions = {
    ignored: ['node_modules', 'cypress', `${output.path}/**/*`]
  },
  notify = true,
  inject = {},
  plugins = [],
  optimization = {}
} = {}) => {
  const isDev = mode === MODE.DEVELOPMENT
  return {
    mode: isDev ? MODE.DEVELOPMENT : MODE.PRODUCTION,
    context,
    entry: flatten(['@babel/polyfill', entry]),
    output: {
      filename: isDev ? 'bundle.js' : 'bundle-[contenthash].js',
      ...output
    },
    watch: isDev && watch !== false,
    watchOptions,
    devtool: isDev ? 'inline-source-map' : undefined,

    module: {
      rules: [
        {
          test: /\.js$/,
          loader: 'babel-loader',
          exclude: /node_modules/
        },
        {
          test: /\.css$/,
          use: ['style-loader', 'css-loader']
        },
        ...(less
          ? [
              {
                test: /\.less$/,
                use: [
                  { loader: 'style-loader' },
                  { loader: 'css-loader' },
                  {
                    loader: 'less-loader',
                    options: { lessOptions }
                  }
                ]
              }
            ]
          : []),
        {
          test: /\.svg$/,
          use: [
            {
              loader: 'babel-loader'
            },
            {
              loader: 'react-svg-loader',
              options: {
                svgo: {
                  plugins: [{ removeTitle: false }, { removeViewBox: false }],
                  floatPrecision: 2
                }
              }
            }
          ]
        },
        {
          // image & font loader (base64 with fallback)
          test: /\.jpe?g$|\.gif$|\.png$|\.wav$|\.mp3|\.woff$|\.woff2$|\.otf$|\.eot$|\.ttf$/,
          type: 'asset',
          generator: {
            filename: '[path][name].[contenthash].[ext]'
          }
        },
        ...rules
      ]
    },

    plugins: [
      new DefinePlugin(
        mapValues(
          {
            __MODE__: mode,
            __GIT_COMMIT__: getGitCommit(),
            ...inject
          },
          ::JSON.stringify
        )
      ),
      ...(html
        ? [
            new HtmlPlugin({
              template: resolve(__dirname, './index.html.ejs'),
              ...htmlOptions
            })
          ]
        : []),
      ...(notify ? [new WebpackNotifierPlugin({ alwaysNotify: true })] : []),
      ...(isDev ? [new LiveReloadPlugin({ appendScriptTag: true })] : [new CompressionPlugin()]),
      ...plugins
    ],

    optimization: {
      minimizer: [
        new TerserPlugin({
          terserOptions: {
            ecma: 6
          }
        })
      ],
      splitChunks: {
        chunks: 'all'
      },
      ...optimization
    }
  }
}
