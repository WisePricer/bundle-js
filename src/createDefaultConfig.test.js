import { resolve } from 'path'
import get from 'lodash.get'
import set from 'lodash.set'

import { createDefaultConfig } from './createDefaultConfig'

const root = resolve('./')
const rebasePath = path => path.replace(root, '/path/to/pkg')
const pathProps = ['context', 'output.path', 'plugins[1].userOptions.template', 'watchOptions.ignored[2]']
const testableDefaultConfig = options => {
  const config = createDefaultConfig(options)
  pathProps.forEach(prop => set(config, prop, rebasePath(get(config, prop))))
  config.plugins.forEach((plugin, idx) => {
    if (plugin.instanceId) set(config, `plugins[${idx}].instanceId`, '__DYNAMIC__')
  })
  set(config, 'plugins[0].definitions.__GIT_COMMIT__', 'abcdef')
  return config
}

describe('createDefaultConfig', () => {
  it('creates default development config', () => {
    expect(testableDefaultConfig()).toMatchSnapshot()
  })

  it('creates default production config', () => {
    expect(testableDefaultConfig({ mode: 'production' })).toMatchSnapshot()
  })

  it('creates a custom config', () => {
    const theme = {
      primaryColor: 'indigo'
    }
    expect(
      testableDefaultConfig({
        context: '/path/to/pkg/client/src',
        entry: ['./app.js', './libs.js'],
        output: {
          path: '/path/to/pkg/build'
        },
        lessOptions: {
          modifyVars: theme
        },
        notify: false,
        inject: {
          __THEME__: theme
        },
        plugins: [{ custom: 'webpack plugin' }]
      })
    ).toMatchSnapshot()
  })
})
