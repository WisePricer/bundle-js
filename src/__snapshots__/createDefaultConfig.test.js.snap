// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createDefaultConfig creates a custom config 1`] = `
Object {
  "context": "/path/to/pkg/client/src",
  "devtool": "inline-source-map",
  "entry": Array [
    "@babel/polyfill",
    "./app.js",
    "./libs.js",
  ],
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "loader": "babel-loader",
        "test": /\\\\\\.js\\$/,
      },
      Object {
        "test": /\\\\\\.css\\$/,
        "use": Array [
          "style-loader",
          "css-loader",
        ],
      },
      Object {
        "test": /\\\\\\.less\\$/,
        "use": Array [
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
          },
          Object {
            "loader": "less-loader",
            "options": Object {
              "lessOptions": Object {
                "modifyVars": Object {
                  "primaryColor": "indigo",
                },
              },
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.svg\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
          },
          Object {
            "loader": "react-svg-loader",
            "options": Object {
              "svgo": Object {
                "floatPrecision": 2,
                "plugins": Array [
                  Object {
                    "removeTitle": false,
                  },
                  Object {
                    "removeViewBox": false,
                  },
                ],
              },
            },
          },
        ],
      },
      Object {
        "generator": Object {
          "filename": "[path][name].[contenthash].[ext]",
        },
        "test": /\\\\\\.jpe\\?g\\$\\|\\\\\\.gif\\$\\|\\\\\\.png\\$\\|\\\\\\.wav\\$\\|\\\\\\.mp3\\|\\\\\\.woff\\$\\|\\\\\\.woff2\\$\\|\\\\\\.otf\\$\\|\\\\\\.eot\\$\\|\\\\\\.ttf\\$/,
        "type": "asset",
      },
    ],
  },
  "optimization": Object {
    "minimizer": Array [
      TerserPlugin {
        "options": Object {
          "exclude": undefined,
          "extractComments": true,
          "include": undefined,
          "minify": undefined,
          "parallel": true,
          "terserOptions": Object {
            "ecma": 6,
          },
          "test": /\\\\\\.\\[cm\\]\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
        },
      },
    ],
    "splitChunks": Object {
      "chunks": "all",
    },
  },
  "output": Object {
    "filename": "bundle.js",
    "path": "/path/to/pkg/build",
  },
  "plugins": Array [
    DefinePlugin {
      "definitions": Object {
        "__GIT_COMMIT__": "abcdef",
        "__MODE__": "\\"development\\"",
        "__THEME__": "{\\"primaryColor\\":\\"indigo\\"}",
      },
    },
    HtmlWebpackPlugin {
      "userOptions": Object {
        "template": "/path/to/pkg/src/index.html.ejs",
      },
      "version": 5,
    },
    LiveReloadPlugin {
      "defaultPort": 35729,
      "infrastructureLogger": null,
      "instanceId": "__DYNAMIC__",
      "isWebpack4": false,
      "lastChildHashes": Array [],
      "lastHash": null,
      "options": Object {
        "appendScriptTag": true,
        "delay": 0,
        "hostname": "\\" + location.hostname + \\"",
        "ignore": null,
        "port": 35729,
        "protocol": "",
        "quiet": false,
        "useSourceHash": false,
      },
      "server": null,
      "sourceHashs": Object {},
      "webpack": null,
    },
    Object {
      "custom": "webpack plugin",
    },
  ],
  "watch": true,
  "watchOptions": Object {
    "ignored": Array [
      "node_modules",
      "cypress",
      "/path/to/pkg/build/**/*",
    ],
  },
}
`;

exports[`createDefaultConfig creates default development config 1`] = `
Object {
  "context": "/path/to/pkg/src",
  "devtool": "inline-source-map",
  "entry": Array [
    "@babel/polyfill",
    "./index.js",
  ],
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "loader": "babel-loader",
        "test": /\\\\\\.js\\$/,
      },
      Object {
        "test": /\\\\\\.css\\$/,
        "use": Array [
          "style-loader",
          "css-loader",
        ],
      },
      Object {
        "test": /\\\\\\.less\\$/,
        "use": Array [
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
          },
          Object {
            "loader": "less-loader",
            "options": Object {
              "lessOptions": Object {},
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.svg\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
          },
          Object {
            "loader": "react-svg-loader",
            "options": Object {
              "svgo": Object {
                "floatPrecision": 2,
                "plugins": Array [
                  Object {
                    "removeTitle": false,
                  },
                  Object {
                    "removeViewBox": false,
                  },
                ],
              },
            },
          },
        ],
      },
      Object {
        "generator": Object {
          "filename": "[path][name].[contenthash].[ext]",
        },
        "test": /\\\\\\.jpe\\?g\\$\\|\\\\\\.gif\\$\\|\\\\\\.png\\$\\|\\\\\\.wav\\$\\|\\\\\\.mp3\\|\\\\\\.woff\\$\\|\\\\\\.woff2\\$\\|\\\\\\.otf\\$\\|\\\\\\.eot\\$\\|\\\\\\.ttf\\$/,
        "type": "asset",
      },
    ],
  },
  "optimization": Object {
    "minimizer": Array [
      TerserPlugin {
        "options": Object {
          "exclude": undefined,
          "extractComments": true,
          "include": undefined,
          "minify": undefined,
          "parallel": true,
          "terserOptions": Object {
            "ecma": 6,
          },
          "test": /\\\\\\.\\[cm\\]\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
        },
      },
    ],
    "splitChunks": Object {
      "chunks": "all",
    },
  },
  "output": Object {
    "filename": "bundle.js",
    "path": "/path/to/pkg/dist",
  },
  "plugins": Array [
    DefinePlugin {
      "definitions": Object {
        "__GIT_COMMIT__": "abcdef",
        "__MODE__": "\\"development\\"",
      },
    },
    HtmlWebpackPlugin {
      "userOptions": Object {
        "template": "/path/to/pkg/src/index.html.ejs",
      },
      "version": 5,
    },
    WebpackNotifierPlugin {
      "isFirstBuild": true,
      "lastBuildSucceeded": false,
      "options": Object {
        "alwaysNotify": true,
      },
    },
    LiveReloadPlugin {
      "defaultPort": 35729,
      "infrastructureLogger": null,
      "instanceId": "__DYNAMIC__",
      "isWebpack4": false,
      "lastChildHashes": Array [],
      "lastHash": null,
      "options": Object {
        "appendScriptTag": true,
        "delay": 0,
        "hostname": "\\" + location.hostname + \\"",
        "ignore": null,
        "port": 35729,
        "protocol": "",
        "quiet": false,
        "useSourceHash": false,
      },
      "server": null,
      "sourceHashs": Object {},
      "webpack": null,
    },
  ],
  "watch": true,
  "watchOptions": Object {
    "ignored": Array [
      "node_modules",
      "cypress",
      "/path/to/pkg/dist/**/*",
    ],
  },
}
`;

exports[`createDefaultConfig creates default production config 1`] = `
Object {
  "context": "/path/to/pkg/src",
  "devtool": undefined,
  "entry": Array [
    "@babel/polyfill",
    "./index.js",
  ],
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "loader": "babel-loader",
        "test": /\\\\\\.js\\$/,
      },
      Object {
        "test": /\\\\\\.css\\$/,
        "use": Array [
          "style-loader",
          "css-loader",
        ],
      },
      Object {
        "test": /\\\\\\.less\\$/,
        "use": Array [
          Object {
            "loader": "style-loader",
          },
          Object {
            "loader": "css-loader",
          },
          Object {
            "loader": "less-loader",
            "options": Object {
              "lessOptions": Object {},
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.svg\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
          },
          Object {
            "loader": "react-svg-loader",
            "options": Object {
              "svgo": Object {
                "floatPrecision": 2,
                "plugins": Array [
                  Object {
                    "removeTitle": false,
                  },
                  Object {
                    "removeViewBox": false,
                  },
                ],
              },
            },
          },
        ],
      },
      Object {
        "generator": Object {
          "filename": "[path][name].[contenthash].[ext]",
        },
        "test": /\\\\\\.jpe\\?g\\$\\|\\\\\\.gif\\$\\|\\\\\\.png\\$\\|\\\\\\.wav\\$\\|\\\\\\.mp3\\|\\\\\\.woff\\$\\|\\\\\\.woff2\\$\\|\\\\\\.otf\\$\\|\\\\\\.eot\\$\\|\\\\\\.ttf\\$/,
        "type": "asset",
      },
    ],
  },
  "optimization": Object {
    "minimizer": Array [
      TerserPlugin {
        "options": Object {
          "exclude": undefined,
          "extractComments": true,
          "include": undefined,
          "minify": undefined,
          "parallel": true,
          "terserOptions": Object {
            "ecma": 6,
          },
          "test": /\\\\\\.\\[cm\\]\\?js\\(\\\\\\?\\.\\*\\)\\?\\$/i,
        },
      },
    ],
    "splitChunks": Object {
      "chunks": "all",
    },
  },
  "output": Object {
    "filename": "bundle-[contenthash].js",
    "path": "/path/to/pkg/dist",
  },
  "plugins": Array [
    DefinePlugin {
      "definitions": Object {
        "__GIT_COMMIT__": "abcdef",
        "__MODE__": "\\"production\\"",
      },
    },
    HtmlWebpackPlugin {
      "userOptions": Object {
        "template": "/path/to/pkg/src/index.html.ejs",
      },
      "version": 5,
    },
    WebpackNotifierPlugin {
      "isFirstBuild": true,
      "lastBuildSucceeded": false,
      "options": Object {
        "alwaysNotify": true,
      },
    },
    CompressionPlugin {
      "algorithm": [Function],
      "options": Object {
        "algorithm": "gzip",
        "compressionOptions": Object {
          "level": 9,
        },
        "deleteOriginalAssets": false,
        "exclude": undefined,
        "filename": "[path][base].gz",
        "include": undefined,
        "minRatio": 0.8,
        "test": undefined,
        "threshold": 0,
      },
    },
  ],
  "watch": false,
  "watchOptions": Object {
    "ignored": Array [
      "node_modules",
      "cypress",
      "/path/to/pkg/dist/**/*",
    ],
  },
}
`;
